manager/allocator/network.go

```diff
@@ -927,8 +1019,77 @@ func (a *Allocator) allocateNode(ctx context.Context, node *api.Node, existingAd
 
 		allocated = true
 	}
+
+	// if we're only initializing existing addresses, we should stop here and
+	// not deallocate anything
+	if existingAddressesOnly {
+		return allocated
+	}
+
+	// now that we've allocated everything new, we have to remove things that
+	// do not belong. we have to do this last because we can easily roll back
+	// attachments we've allocated if something goes wrong by freeing them, but
+	// we can't roll back deallocating attachments by reacquiring them.
+	for i, attach := range node.Attachments {
+		// for every attachment, go through every network. if the attachment
+		// belongs to one of the networks, then go to the next attachment. if
+		// no network matches
+		attachmentBelongs := false
+		for _, network := range networks {
+			if network.ID == attach.Network.ID {
+				attachmentBelongs = true
+				break
+			}
+		}
+		if !attachmentBelongs {
+			// free the attachment and remove it from the node's attachments by
+			// re-slicing
+			if err := a.netCtx.nwkAllocator.DeallocateAttachment(node, attach); err != nil {
+				// if deallocation fails, there's nothing we can do besides log
+				// an error and keep going
+				log.G(ctx).WithError(err).Errorf(
+					"error deallocating attachment for network %v on node %v",
+					attach.Network.ID, node.ID,
+				)
+			}
+			node.Attachments = append(node.Attachments[0:i], node.Attachments[i+1:]...)
```

Comment:
May be being stupid here, but this seems like it would not work correctly.   We are in a for-loop over the attachments.  If we are on attachment `i` and we remove it from the slice, then attachment `i+1` now occupies position `i`.  But the loop is next going to inspect index `i+1` which was previously `i+2`.  In other words we just skipped the attachment *originally* at `i+1` before `i` was removed.   Did I get that right?

Reply:
huh... maybe. i'm unsure how this works actually. good catch.